File: DSA/src/Questions/DP/LongestCommonSubsequence.java
package Questions.DP;

import java.util.Arrays;

public class LongestCommonSubsequence {
    static int lcsUtil(String s1, String s2, int ind1, int ind2, int[][] dp) {
        // Base case: If either of the strings reaches the end, return 0
        if (ind1 < 0 || ind2 < 0)
            return 0;

        // If the result for this subproblem has already been calculated, return it
        if (dp[ind1][ind2] != -1)
            return dp[ind1][ind2];

        // If the characters at the current indices are the same, increment the LCS
        // length
        if (s1.charAt(ind1) == s2.charAt(ind2))
            return dp[ind1][ind2] = 1 + lcsUtil(s1, s2, ind1 - 1, ind2 - 1, dp);

        // If the characters are different, choose the maximum LCS length by either
        // skipping a character in s1 or skipping a character in s2
        else
            return dp[ind1][ind2] = Math.max(lcsUtil(s1, s2, ind1, ind2 - 1, dp),
                    lcsUtil(s1, s2, ind1 - 1, ind2, dp));
    }

    // Function to find the length of the Longest Common Subsequence (LCS)
    static int lcsRecursive(String s1, String s2) {
        int n = s1.length();
        int m = s2.length();

        // Create a 2D array to store results of subproblems
        int[][] dp = new int[n][m];

        // Initialize the dp array with -1 to indicate that subproblems are not solved
        // yet
        for (int[] rows : dp)
            Arrays.fill(rows, -1);

        // Call the recursive function to find the LCS length
        return lcsUtil(s1, s2, n - 1, m - 1, dp);
    }

    // Time Complexity: O(N*M)

    // Reason: There are N*M states therefore at max ‘N*M’ new problems will be
    // solved.

    // Space Complexity: O(N*M) + O(N+M)

    // Reason: We are using an auxiliary recursion stack space(O(N+M)) (see the
    // recursive tree, in the worst case, we will go till N+M calls at a time) and a
    // 2D array ( O(N*M)).

    // Function to find the length of the Longest Common Subsequence (LCS)
    static int lcsTabulation(String s1, String s2) {
        int n = s1.length();
        int m = s2.length();

        // Each cell (i,j) in the dp array indicates the length of the LCS for s1[0:i-1]
        // and s2[0:j-1]
        int[][] dp = new int[n + 1][m + 1];

        // Initialize the dp array with -1 to indicate that subproblems are not solved
        // yet
        for (int[] row : dp)
            Arrays.fill(row, -1);

        // Initialize the first row and first column with zeros since LCS with an empty
        // string is zero
        for (int i = 0; i <= n; i++) {
            dp[i][0] = 0;
        }
        for (int i = 0; i <= m; i++) {
            dp[0][i] = 0;
        }

        // Fill the dp array using dynamic programming
        for (int ind1 = 1; ind1 <= n; ind1++) {
            for (int ind2 = 1; ind2 <= m; ind2++) {
                // If the characters at the current indices are the same, increment the LCS
                // length
                if (s1.charAt(ind1 - 1) == s2.charAt(ind2 - 1))
                    dp[ind1][ind2] = 1 + dp[ind1 - 1][ind2 - 1];
                // If the characters are different, choose the maximum LCS length by either
                // excluding a character in s1 or excluding a character in s2
                else
                    dp[ind1][ind2] = Math.max(dp[ind1 - 1][ind2], dp[ind1][ind2 - 1]);
            }
        }

        return dp[n][m]; // Return the length of the Longest Common Subsequence (LCS)
    }
    // Time Complexity: O(N*M)

    // Reason: There are two nested loops

    // Space Complexity: O(N*M)

    // Reason: We are using an external array of size ‘N*M)’. Stack Space is
    // eliminated.

    // Function to find the length of the Longest Common Subsequence (LCS)
    static int lcsOptimized(String s1, String s2) {
        int n = s1.length();
        int m = s2.length();

        // Create arrays to store the LCS lengths
        int[] prev = new int[m + 1];
        int[] cur = new int[m + 1];

        // Iterate through the strings and calculate LCS lengths
        for (int ind1 = 1; ind1 <= n; ind1++) {
            for (int ind2 = 1; ind2 <= m; ind2++) {
                // If the characters at the current indices are the same, increment the LCS
                // length
                if (s1.charAt(ind1 - 1) == s2.charAt(ind2 - 1))
                    cur[ind2] = 1 + prev[ind2 - 1];
                // If the characters are different, choose the maximum LCS length by either
                // excluding a character in s1 or excluding a character in s2
                else
                    cur[ind2] = Math.max(prev[ind2], cur[ind2 - 1]);
            }
            // Update the 'prev' array to the values of 'cur' for the next iteration
            prev = (int[]) (cur.clone());
        }

        return prev[m]; // Return the length of the Longest Common Subsequence (LCS)
    }
    // Time Complexity: O(N*M)

    // Reason: There are two nested loops.

    // Space Complexity: O(M)

    // Reason: We are using an external array of size ‘M+1’ to store only two rows.

    public static void main(String[] args) {
        String s1 = "AGGTAB";
        String s2 = "GXTXAYB";

        System.out.println(lcsRecursive(s1, s2)); // Output: 4
        System.out.println(lcsTabulation(s1, s2)); // Output: 4
        System.out.println(lcsOptimized(s1, s2)); // Output: 4
    }

}

File: DSA/Python/Questions/DP/longest_common_subsequence.py
"""
# Question: Longest Common Subsequence
# Link: https://leetcode.com/problems/longest-common-subsequence/

# Problem Statement:
# Given two strings text1 and text2, return the length of their longest common subsequence.
# A subsequence is a sequence that can be derived from another sequence by deleting some or
# no elements without changing the order of the remaining elements.

# Example:
# Input: text1 = "AGGTAB", text2 = "GXTXAYB"
# Output: 4
# Explanation: The longest common subsequence is "GTAB"
"""

# 1. Recursive Solution with Memoization
"""
Algorithm:
1. Compare characters from both strings:
   - If match, include character and move both pointers
   - If not match, try excluding character from either string
2. Use memoization to store results of subproblems

Time Complexity: O(n*m)
Space Complexity: O(n*m) + O(n+m) recursion stack
"""


def lcs_recursive(s1: str, s2: str) -> int:
    n, m = len(s1), len(s2)
    # Initialize dp array for memoization
    dp = [[-1] * m for _ in range(n)]

    def solve(i: int, j: int) -> int:
        # Base case: if either string is exhausted
        if i < 0 or j < 0:
            return 0

        # Return memoized result if available
        if dp[i][j] != -1:
            return dp[i][j]

        # If characters match
        if s1[i] == s2[j]:
            dp[i][j] = 1 + solve(i - 1, j - 1)
        else:
            # Try excluding character from either string
            dp[i][j] = max(solve(i - 1, j), solve(i, j - 1))

        return dp[i][j]

    return solve(n - 1, m - 1)


# 2. Tabulation Solution
"""
Algorithm:
1. Create 2D dp array of size (n+1)*(m+1)
2. Fill dp table bottom-up:
   - If characters match, add 1 to diagonal value
   - If not match, take max of left and top values

Time Complexity: O(n*m)
Space Complexity: O(n*m)
"""


def lcs_tabulation(s1: str, s2: str) -> int:
    n, m = len(s1), len(s2)
    # Create dp array with base cases
    dp = [[0] * (m + 1) for _ in range(n + 1)]

    # Fill dp table
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            if s1[i - 1] == s2[j - 1]:
                dp[i][j] = 1 + dp[i - 1][j - 1]
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])

    return dp[n][m]


# 3. Space Optimized Solution
"""
Algorithm:
1. Use two arrays instead of 2D array
2. Previous row and current row sufficient
3. Update arrays in alternating fashion

Time Complexity: O(n*m)
Space Complexity: O(m)
"""


def lcs_optimized(s1: str, s2: str) -> int:
    n, m = len(s1), len(s2)
    # Initialize previous row
    prev = [0] * (m + 1)

    # Process each character of s1
    for i in range(1, n + 1):
        # Current row array
        curr = [0] * (m + 1)

        # Fill current row
        for j in range(1, m + 1):
            if s1[i - 1] == s2[j - 1]:
                curr[j] = 1 + prev[j - 1]
            else:
                curr[j] = max(prev[j], curr[j - 1])

        # Update previous row
        prev = curr[:]

    return prev[m]


def main():
    # Test cases
    test_cases = [
        ("AGGTAB", "GXTXAYB"),  # Regular case
        ("ABCDGH", "AEDFHR"),  # Multiple common subsequences
        ("ABC", "DEF"),  # No common subsequence
        ("AAA", "AAA"),  # Same strings
        ("", "ABC"),  # Empty string
    ]

    for i, (s1, s2) in enumerate(test_cases):
        print(f"\nTest Case {i + 1}: s1 = '{s1}', s2 = '{s2}'")
        print(f"Recursive Solution: {lcs_recursive(s1, s2)}")
        print(f"Tabulation Solution: {lcs_tabulation(s1, s2)}")
        print(f"Optimized Solution: {lcs_optimized(s1, s2)}")


if __name__ == "__main__":
    main()
